builtin/logical/totp/path_roles.go

```diff
@@ -0,0 +1,342 @@
+				"digits":       role.Digits,
+			},
+		}, nil
+	}
+}
+
+func (b *backend) pathRoleList(
+	req *logical.Request, d *framework.FieldData) (*logical.Response, error) {
+	entries, err := req.Storage.List("key/")
+	if err != nil {
+		return nil, err
+	}
+
+	return logical.ListResponse(entries), nil
+}
+
+func (b *backend) pathRoleCreate(
+	req *logical.Request, data *framework.FieldData) (*logical.Response, error) {
+	name := data.Get("name").(string)
+	generate := data.Get("generate").(bool)
+	key_string := data.Get("key").(string)
+	issuer := data.Get("issuer").(string)
+	account_name := data.Get("account_name").(string)
+	period := data.Get("period").(int)
+	algorithm := data.Get("algorithm").(string)
+	digits := data.Get("digits").(int)
+	skew := data.Get("skew").(int)
+	qr_size := data.Get("qr_size").(int)
+	key_size := data.Get("key_size").(int)
+
+	// Translate digits and algorithm to a format the totp library understands
+	var role_digits otplib.Digits
+	switch digits {
+	case 6:
+		role_digits = otplib.DigitsSix
+	case 8:
+		role_digits = otplib.DigitsEight
+	default:
+		return logical.ErrorResponse("The digit value can only be 6 or 8."), nil
+	}
+
+	var role_algorithm otplib.Algorithm
+	switch algorithm {
+	case "SHA1":
+		role_algorithm = otplib.AlgorithmSHA1
+	case "SHA256":
+		role_algorithm = otplib.AlgorithmSHA256
+	case "SHA512":
+		role_algorithm = otplib.AlgorithmSHA512
+	default:
+		return logical.ErrorResponse("The algorithm value is not valid."), nil
+	}
+
+	// Enforce input value requirements
+	if period <= 0 {
+		return logical.ErrorResponse("The period value must be greater than zero."), nil
+	}
+
+	if skew < 0 {
+		return logical.ErrorResponse("The skew value must be greater than zero."), nil
+	}
+
+	if qr_size <= 0 {
+		return logical.ErrorResponse("The qr_size value must be greater than zero."), nil
+	}
+
+	if key_size <= 0 {
+		return logical.ErrorResponse("The key_size value must be greater than zero."), nil
+	}
+
+	// If the key is generated, Account Name and Issuer are required.
+	if generate {
+		if account_name == "" {
+			return logical.ErrorResponse("The account_name value is required for generated keys."), nil
+		}
+
+		if issuer == "" {
+			return logical.ErrorResponse("The issuer value is required for generated keys."), nil
+		}
+	}
+
+	// Period, Skew and Key Size need to be unsigned ints
+	uint_period := uint(period)
+	uint_skew := uint(skew)
+	uint_key_size := uint(key_size)
+
+	url := ""
+
+	switch generate {
+	case true:
+		// Generate a new key
+		key_object, err := totplib.Generate(totplib.GenerateOpts{
+			Issuer:      issuer,
+			AccountName: account_name,
+			Period:      uint_period,
+			Digits:      role_digits,
+			Algorithm:   role_algorithm,
+			SecretSize:  uint_key_size,
+		})
+
+		if err != nil {
+			return logical.ErrorResponse("An error occured while generating a key."), nil
+		}
+
+		url = key_object.String()
+		key_string = key_object.Secret()
+	case false:
+		if key_string == "" {
```

Comment:
Should add support for parsing a URL in this case so that the issuer/account name/etc. can be automatically determined.